### PR TITLE
Add I18n phrase

### PR DESCRIPTION
**Pull request by fvovan**
@njam, @tomaszdurka please review

### CODE REVIEW

**Comment by tomaszdurka** [comment]
Looks good, couple of things though.

- We should fine a bit better name for this class :)
- You don't need that docu code about instance type in translate smarty helper. Having `if (... instanceof... )` will automatically consider it as instance inside that if.
- Let's rename `$string`, `$params` to `$phrase`, `$variables` as is in several places.

Cheers!

**Comment by fvovan** [comment]
what about 'Localized' ?

**Comment by njam** [comment]
Isn't "I18n" more of a general concept?

In general we're trying to move common files into "module" directories at the root level of "library". So maybe `CM_Translate_Phrase` or `CM_I18n_Phrase`?

**Comment by fvovan** [comment]
I opt for CM_Translate_Phrase 
and what is corresponding module for it?  Or should I create new one with only single class?

**Comment by njam** [comment]
There's no module for this topic yet.
@tomaszdurka what's your opinion?

File: library/CM/SmartyPlugins/function.translateVariable.php [hunk 1]
```diff
@@ -7,5 +7,10 @@ function smarty_function_translateVariable($params, Smarty_Internal_Template $te
     $key = $params['key'];
     unset($params['key']);
 
-    return $render->getTranslation($key, $params);
+    if ($key instanceof CM_I18n) {
```
Comment (tomaszdurka) [inline library/CM/SmartyPlugins/function.translateVariable.php]: Add check about `$params` size and throw to let the developer know that those params will be totally ignored.

File: library/CM/I18n.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+class CM_I18n {
```
Comment (tomaszdurka) [inline library/CM/I18n.php]: CM_I18n is great name for module containing all the language/translation related classes.
It's not the best moment to move all those classes now.

Let's only rename this class to CM_I18n_Phrase and make it first one being properly placed in the new module.

**Comment by tomaszdurka** [comment]
@fvovan Review is done, thanks!

File: library/CM/I18n/Phrase.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+class CM_I18n_Phrase {
+
+    protected $phrase, $variables;
```
Comment (tomaszdurka) [inline library/CM/I18n/Phrase.php]: Add docu.

File: tests/library/CM/SmartyPlugins/function.translateVariableTest.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+require_once CM_Util::getModulePath('CM') . 'library/CM/SmartyPlugins/prefilter.translate.php';
+
+class smarty_function_translateVariableTest extends CMTest_TestCase {
+
+    /**
+     * @expectedException        InvalidArgumentException
+     * @expectedExceptionMessage Passed params will be ignored as you provided CM_I18n_Phrase object
+     */
+    public function testTranslateObject() {
```
Comment (tomaszdurka) [inline tests/library/CM/SmartyPlugins/function.translateVariableTest.php]: rename to testTranslatePhrase

File: library/CM/I18n/Phrase.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+    /**
+     * @param string $phrase
+     * @param array  $variables
+     */
+    public function __construct($phrase, array $variables) {
+        $this->phrase = $phrase;
```
Comment (tomaszdurka) [inline library/CM/I18n/Phrase.php]: Cast to string.

File: library/CM/I18n/Phrase.php (hunk 3 above)
Comment (tomaszdurka) [inline library/CM/I18n/Phrase.php]: Probably use `string[]` for variables.

File: tests/library/CM/SmartyPlugins/function.translateVariableTest.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+            $this->assertSame(['bar' => 3], $params);
+        });
+        /** @var CM_Frontend_Render $render */
+
+        $object = new CM_I18n_Phrase('Bar value is {$bar}', ['bar' => 3]);
+        $render->parseTemplateContent('{translateVariable key=$foo}', ['foo' => $object]);
+        $this->assertSame(1, $getTranslationMethod->getCallCount());
+
+        $render->parseTemplateContent('{translateVariable key=$foo more=one}', ['foo' => $object]);
+        $this->assertSame(0, $getTranslationMethod->getCallCount());
```
Comment (tomaszdurka) [inline tests/library/CM/SmartyPlugins/function.translateVariableTest.php]: Adjust to use `$this->catchException()` as discussed.

**Comment by fvovan** [comment]
@njam, please doublecheck

**Comment by njam** [comment]
- CM_I18n_Phrase: Non-public members should be prefixed with `_`
- CM_I18n_Phrase::__construct: Docu for $variables could be `string[]`
- CM_I18n_Phrase: Add simple unit test
- smarty_function_translateVariable: How about also creating an `CM_I18n_Phrase` instance if we don't pass one? Might be simpler in the end?

**Comment by fvovan** [comment]
I added tests for CM_I18n, but now one assertion is duplicated with function.translateVariableTest.php
On the one hand in smarty tests we should check only correct CM_I18n_Phrase state during translate invocation.
On the other hand we can't do it because its' properties are protected.
@njam what do you think?

**Comment by njam** [comment]
- CM_I18n_PhraseTest: formatting

Rest looks good. I think having duplication in tests is okay.

**Comment by tomaszdurka** [comment]
Looks good, merging.

**Comment by tomaszdurka** [comment]
@fvovan To use in in external, dependant projects you need to tag it. Could you do it? If you have any problems please let me know.

**Comment by fvovan** [comment]
@tomaszdurka,  it's done
1.126.0

**Pull request merged**